src/workerd/api/node/i18n.c++

```diff
@@ -0,0 +1,258 @@
+                             const_cast<const UChar*>(sourcebuf.begin()), length_in_chars, &status);
+
+  if (U_SUCCESS(status)) {
+    return destbuf.slice(0, len).attach(kj::mv(destbuf));
+  }
+
+  return kj::none;
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeUcs2FromUtf8(kj::ArrayPtr<kj::byte> source,
+                                                     Encoding fromEncoding, Encoding toEncoding) {
+  size_t expected_utf16_length =
+      simdutf::utf16_length_from_utf8(source.asChars().begin(), source.size());
+  // Workers are limited to 128MB so this isn't actually a realistic concern, but sanity check.
+  JSG_REQUIRE(expected_utf16_length <= ISOLATE_LIMIT, Error,
+              "Expected UTF-16le length is too large to transcode");
+  auto destbuf = kj::heapArray<UChar>(expected_utf16_length);
+
+  size_t actual_length =
+      simdutf::convert_utf8_to_utf16le(source.asChars().begin(), source.size(), destbuf.begin());
+  JSG_REQUIRE(actual_length == expected_utf16_length, Error, "Expected UTF16 length mismatch");
+
+  // simdutf returns 0 for invalid UTF-8 value.
+  if (actual_length == 0) {
+    return kj::none;
+  }
+
+  return destbuf.asBytes().attach(kj::mv(destbuf));
+}
+
+kj::Maybe<kj::Array<kj::byte>> TranscodeUtf8FromUcs2(kj::ArrayPtr<kj::byte> source,
+                                                     Encoding fromEncoding, Encoding toEncoding) {
+  JSG_REQUIRE(source.size() % 2 == 0, Error, "UTF-16le input size should be multiple of 2");
+  auto utf16_input =
+      kj::arrayPtr<char16_t>(reinterpret_cast<char16_t*>(source.begin()), source.size() / 2);
+  size_t expected_utf8_length =
+      simdutf::utf8_length_from_utf16le(utf16_input.begin(), utf16_input.size());
+
+  // Workers are limited to 128MB so this isn't actually a realistic concern, but sanity check.
+  JSG_REQUIRE(expected_utf8_length <= ISOLATE_LIMIT, Error,
+              "Expected UTF-8 length is too large to transcode");
+
+  auto destbuf = kj::heapArray<kj::byte>(expected_utf8_length);
+
+  size_t actual_length = simdutf::convert_utf16le_to_utf8(utf16_input.begin(), utf16_input.size(),
+                                                          destbuf.asChars().begin());
+  JSG_REQUIRE(actual_length == expected_utf8_length, Error, "Expected UTF8 length mismatch");
+
+  // simdutf returns 0 for invalid UTF-8 value.
+  if (actual_length == 0) {
+    return kj::none;
+  }
+
+  return destbuf.asBytes().attach(kj::mv(destbuf));
+}
+
+} // namespace
+
+Converter::Converter(Encoding encoding, kj::StringPtr substitude) {
+  UErrorCode status = U_ZERO_ERROR;
+  auto name = getEncodingName(encoding);
+  auto conv = ucnv_open(name, &status);
+  JSG_REQUIRE(U_SUCCESS(status), Error, "Failed to initialize converter");
+  conv_ = kj::Own<UConverter>(conv, ConverterDisposer::INSTANCE);
+  setSubstitudeChars(substitude);
+}
+
+Converter::Converter(UConverter* converter, kj::StringPtr substitude)
```

Comment:
Is this constructor required? 
If so, can we not assume converter would need to be NONNULL and move the `KJ_ASSERT_NONNULL(conv_.get());` check from all the methods to this constructor so that we check that at creation rather than on every method call?

Reply:
Actually it is not needed.